### PR TITLE
Fixes #106142.

### LINKED ISSUE #106142: JIT: Bad codegen with `Avx512F.VL.RotateLeftVariable` and `Avx512F.VL.RotateRightVariable` on win-x86
```csharp
// Generated by Fuzzlyn v2.2 on 2024-08-06 15:26:23
// Run on X86 Windows
// Seed: 17786759086488605032-vectort,vector128,vector256,x86aes,x86avx,x86avx2,x86avx512bw,x86avx512bwvl,x86avx512cd,x86avx512cdvl,x86avx512dq,x86avx512dqvl,x86avx512f,x86avx512fvl,x86bmi1,x86bmi2,x86fma,x86lzcnt,x86pclmulqdq,x86popcnt,x86sse,x86sse2,x86sse3,x86sse41,x86sse42,x86ssse3,x86x86base
// Reduced from 186.1 KiB to 0.4 KiB in 00:13:36
// Debug: Outputs <4611686018427387904, 0, 0, 0>
// Release: Outputs <0, 0, 0, 0>
using System;
using System.Runtime.CompilerServices;
using System.Numerics;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;

public class Program
{
    public static void Main()
    {
        var vr7 = Vector256.Create(1, 0, 0, 0);
        var vr8 = Vector256.CreateScalar(18446744073709551614UL);
        var vr9 = Avx512F.VL.RotateLeftVariable(vr7, vr8);
        System.Console.WriteLine(vr9);
    }
}
```

```csharp
// Generated by Fuzzlyn v2.2 on 2024-08-06 15:23:25
// Run on X86 Windows
// Seed: 4121896490095390139-vectort,vector128,vector256,x86aes,x86avx,x86avx2,x86avx512bw,x86avx512bwvl,x86avx512cd,x86avx512cdvl,x86avx512dq,x86avx512dqvl,x86avx512f,x86avx512fvl,x86bmi1,x86bmi2,x86fma,x86lzcnt,x86pclmulqdq,x86popcnt,x86sse,x86sse2,x86sse3,x86sse41,x86sse42,x86ssse3,x86x86base
// Reduced from 31.4 KiB to 0.4 KiB in 00:01:15
// Debug: Outputs <9007199254740992, 1, 1, 1>
// Release: Outputs <0, 1, 1, 1>
using System;
using System.Runtime.CompilerServices;
using System.Numerics;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.X86;

public class Program
{
    public static void Main()
    {
        var vr2 = Vector256.Create<long>(1);
        var vr3 = Vector256.CreateScalar(9945469575827037067UL);
        Vector256<long> vr4 = Avx512F.VL.RotateRightVariable(vr2, vr3);
        System.Console.WriteLine(vr4);
    }
}
```
cc @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by DeepakRajendrakumaran**
This Fixes #106142 

Rotate requires the count to be normalized depending on base type. This was already done with https://github.com/dotnet/runtime/commit/f251b170fb1d1be8552e7e1f3238aca9bcc6bb7a but there was a behavioral mismatch between x86 and x64 with the mod(%) operator

### CODE REVIEW

**Comment by JulieLeeMSFT** [comment]
@tannergooding @jakobbotsch , please review the code before RC1 snap.
Cc @dotnet/jit-contrib .

File: src/coreclr/jit/simd.h [hunk 1]
```diff
@@ -802,17 +802,33 @@ TBase EvaluateBinaryScalarSpecialized(genTreeOps oper, TBase arg0, TBase arg1)
...
-            arg1 %= (sizeof(TBase) * BITS_PER_BYTE);
+            if ((arg1 < 0) || (arg1 >= (sizeof(TBase) * BITS_PER_BYTE)))
+            {
+                // EvaluateBinaryScalarRSZ allows overshifting and treats
+                // it as zeroing.
+                // But ROL ensures the rotateAmount is masked
+                // to be within range, so we pre-calculates this.
+                unsigned rotateCountMask = (sizeof(TBase) * BITS_PER_BYTE) - 1;
+                arg1 &= rotateCountMask;
+            }
```
Comment (jakobbotsch) [inline src/coreclr/jit/simd.h:813]: Are the ifs needed?
```suggestion
            arg1 &= (sizeof(TBase) * BITS_PER_BYTE) - 1;
```

**Review by jakobbotsch — COMMENTED** [review]
Can you add the test?

**Review by jakobbotsch — APPROVED** [review]

**Comment by JulieLeeMSFT** [comment]
Thanks for the work. If tests pass, I will merge today.

**Review by tannergooding — APPROVED** [review]

File: src/coreclr/jit/simd.h [hunk 2]
```diff
@@ -802,17 +802,27 @@ TBase EvaluateBinaryScalarSpecialized(genTreeOps oper, TBase arg0, TBase arg1)
         case GT_ROL:
         {
             // Normalize the "rotate by" value
-            arg1 %= (sizeof(TBase) * BITS_PER_BYTE);
```
[the remark this answers is not in the document]
Comment (DeepakRajendrakumaran) [reply src/coreclr/jit/simd.h:805]: Yep. It was different for negative inputs. For eg. one of the test case has the rotate count show up as -2 in here.

**Pull request merged**